src/training/graph_group.h

```diff
@@ -83,22 +85,20 @@ class AsyncGraphGroup : public GraphGroup {
         opt_->update(graphs_[0]);
       }
       else {
-        std::lock_guard<std::mutex> guard(sync_);
         // add instead of copy?
-        ThreadPool pool(devices_.size());
         std::vector< std::future<void> > results;
         int pos = 0;
         for (int idx = 0; idx < devices_.size(); idx++) {
-            results.push_back( std::async (std::launch::async, [=](int idx, int pos) {
+          std::thread([=](int idx, int pos) {
+            //individual mutex per-shard
+            std::lock_guard<std::mutex> guard( shardSync_[idx] );
             grads_[idx]->copyFrom( newGrads->subtensor(pos , grads_[idx]->size() ) );
             shardOpt_[idx]->update(params_[idx], grads_[idx]);
-          } , idx, pos));  
 
-          pos += shardSize_;
-        }
+            cudaStreamSynchronize(0);
+          } , idx, pos).detach();  
```

Comment:
It's faster, but I think the issue with worse convergence is back. I think you need to join the threads. 

```
std::vector<std::thread> threads;
...
threads.emplace_back([]() { ... })
...
for(auto&& t : threads)
  t.join();
```